scipy/linalg/basic.py

```diff
@@ -1617,3 +1599,202 @@ def matrix_balance(A, permute=True, scale=True, separate=False,
+    implementations in Python.
+
+    .. versionadded:: 1.5.0
+
+    Examples
+    --------
+    Multiply the Toeplitz matrix T with matrix b::
+
+            [ 1 -1 -2 -3]       [1 10]
+        T = [ 3  1 -1 -2]   b = [2 11]
+            [ 6  3  1 -1]       [2 11]
+            [10  6  3  1]       [5 19]
+
+    To specify the Toeplitz matrix, only the first column and the first
+    row are needed.
+
+    >>> c = np.array([1, 3, 6, 10])    # First column of T
+    >>> r = np.array([1, -1, -2, -3])  # First row of T
+    >>> b = np.array([[1, 10], [2, 11], [2, 11], [5, 19]])
+
+    >>> from scipy.linalg import toeplitz, matmul_toeplitz
+    >>> matmul_toeplitz((c, r), b)
+    array([[-20., -80.], [ -7.,  -8.], [  9.,  85.], [ 33., 218.]])
+
+    Check the result by creating the full Toeplitz matrix and
+    multiplying it by ``b``.
+
+    >>> toeplitz(c, r) @ b
+    array([[-20, -80], [ -7,  -8], [  9,  85], [ 33, 218]])
+
+    The full matrix is never formed explicitly, so this routine
+    is suitable for very large Toeplitz matrices.
+
+    >>> n = 1000000
+    >>> matmul_toeplitz([1] + [0]*(n-1), np.ones(n))
+    array([1., 1., 1., ..., 1., 1., 1.])
+
+    References
+    ----------
+    .. [1] : Jacob R Gardner, Geoff Pleiss, David Bindel, Kilian
+    Q Weinberger, Andrew Gordon Wilson, "GPyTorch: Blackbox Matrix-Matrix
+    Gaussian Process Inference with GPU Acceleration" with contributions
+    from Max Balandat and Ruihan Wu. Available online:
+    https://github.com/cornellius-gp/gpytorch
+
+    .. [2] : J. Demmel, P. Koev, and X. Li, "A Brief Survey of Direct Linear
+    Solvers". In Z. Bai, J. Demmel, J. Dongarra, A. Ruhe, and H. van der
+    Vorst, editors. Templates for the Solution of Algebraic Eigenvalue
+    Problems: A Practical Guide. SIAM, Philadelphia, 2000. Available online:
+    http://www.netlib.org/utk/people/JackDongarra/etemplates/node384.html
+
+    .. [3] : R. Scheibler, E. Bezzam, I. Dokmanić, Pyroomacoustics: A Python
+    package for audio room simulations and array processing algorithms,
+    Proc. IEEE ICASSP, Calgary, CA, 2018.
+    https://github.com/LCAV/pyroomacoustics/blob/pypi-release/pyroomacoustics
+    /adaptive/util.py
+
+    .. [4] : Maranò S, Edwards B, Ferrari G and Fäh D (2017), "Fitting
+    Earthquake Spectra: Colored Noise and Incomplete Data", Bulletin of
+    the Seismological Society of America., January, 2017. Vol. 107(1),
+    pp. 276-291. Implementation available online at:
+    http://mercalli.ethz.ch/~marra/publications/2017_fitting_earthquake_spect
+    ra_colored_noise_and_incomplete_data/ToeplitzOperations.py
+
+    """
+
+    r, c, b, dtype, b_shape = _validate_args_for_toeplitz_ops(
+        c_or_cr, b, check_finite, keep_b_shape=False)
+    n, m = b.shape
+
+    embedded_col = np.concatenate((c, r[-1:0:-1]))
+    padded_b = np.vstack((b, np.zeros((n - 1, m))))
+
+    fft_mat = np.fft.fft(embedded_col, axis=0).reshape(-1, 1)
+    fft_b = np.fft.fft(padded_b, axis=0)
```

Comment:
```suggestion
    fft_b = np.fft.fft(b, axis=0, n=2*n-1)
```
`fft` can handle the zero-padding for you.

Reply:
Changed :)